tests/reg_tests/test_volumes.py

```diff
@@ -2177,18 +2158,53 @@ def regression_test(self, handler, solve=False):
         X[:, :, :, 2] = data[2 * nval : 3 * nval].reshape((9, 7, 11), order="f")
 
-        # Testing Volume with ku, kv, kw, nCtlu, nCtlv, nCtlw
+        def uniformKnots(N, k):
+            """Simple function to generate N uniform knots of
+            order k"""
+
+            knots = np.zeros(N + k)
+            knots[0 : k - 1] = 0.0
+            knots[-k:] = 1.0
+            knots[k - 1 : -k + 1] = np.linspace(0, 1, N - k + 2)
+
+            return knots
+
+        # Testing Volume with ku, kv, kw
         for ku in [2, 4]:
             for kv in [3, 4]:
                 for kw in [2, 4]:
-                    for nCtlu in [4, 6]:
-                        for nCtlv in [8]:
-                            for nCtlw in [5, 7]:
-                                test_name = "volume with ku={}, kv={}, kw={}, nCtlu={}, nCtlv={}, nCtlw={}".format(
-                                    ku, kv, kw, nCtlu, nCtlv, nCtlw
-                                )
-                                volume = Volume(X=X, ku=ku, kv=kv, kw=kw, nCtlu=nCtlu, nCtlv=nCtlv, nCtlw=nCtlw)
-                                volume.recompute()
-                                run_volume_test(volume, handler, test_name)
-                                run_project_test(volume, handler, test_name)
-                                io_test(volume, handler)
+
+                    # get the knot vectors to initialize the volume using X
+                    tu = uniformKnots(X.shape[0], ku)
+                    tv = uniformKnots(X.shape[1], kv)
+                    tw = uniformKnots(X.shape[2], kw)
+
+                    volume = Volume(coef=X, ku=ku, kv=kv, kw=kw, tu=tu, tv=tv, tw=tw)
+
+                    # generate test points that are inside the volume by interpolating X
+                    n_interp = 5
+                    test_pts = np.zeros((n_interp**3, 3))
+                    ipt = 0
+                    for ifrac in np.linspace(0.0, 1.0, n_interp):
+                        for jfrac in np.linspace(0.0, 1.0, n_interp):
+                            for kfrac in np.linspace(0.0, 1.0, n_interp):
+                                test_pts[ipt, :] = volume.getValue(ifrac, jfrac, kfrac)
+                                ipt += 1
+
+                    test_name = f"volume with ku={ku}, kv={kv}, kw={kw}"
+                    io_test(volume, handler)
+                    run_volume_test(volume, handler, test_name)
+
+                    test_name = f"points inside volume with ku={ku}, kv={kv}, kw={kw}"
+                    run_project_test(volume, handler, test_name, pts=test_pts)
+
+                    # also test a few points outside
+                    test_name = f"points outside volume with ku={ku}, kv={kv}, kw={kw}"
+                    test_pts_outside = np.array(
+                        [
+                            [1.0, 1.0, 1.0],
+                            [0.15, 0.017, 0.39],
+                            [-0.120510712442222084, 0.0551918480620798083, -0.1],
```

Comment:
Can you add some comments on which case each of these points tests?

Reply:
I changed the points here to explicitly test the face, edge, and corner cases.